[CDAP-6357] Added scure store macro function to DefaultMacroEvaluator

knazir opened the pull request:
Added the secure store macro function to DefaultMacroEvaluator. Syntax:

```
${secure(key)}
```

Commit:
[CDAP-6357] Added secure store macro function to DefaultMacroEvaluator

gsps1 commented:
is it possible to add a test case for this ?

knazir commented:
Yeah I'm adding a test case right now

chtyim commented on cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java:
```diff
@@ -73,9 +82,15 @@ public String evaluate(String macroFunction, String... arguments) throws Invalid
+      try {
+        return new String(secureStore.get(namespace, arguments[0]).get());
+      } catch (IOException e) {
+        throw new InvalidMacroException(e.getMessage());
```
Should include the original exception as the cause and with message that is suitable for this context. E.g.

`new InvalidMacroException("Failed to resolve macro '" + macroFunction + "(" + arguments[0] + ")'", e);`

chtyim commented:
Just one comment about the exception.

knazir commented:
Addressed the comment, thanks Terence. Still trying to figure out a unit test with access to both the secure store and DefaultMacroEvaluator or a suitable alternative...

knazir commented:
Ok got a unit test... looks like it's working. Any comments on exposing secure store in the TestBase API? It may be useful for other tests as well.

gsps1 commented on cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java:
```diff
@@ -32,12 +34,14 @@
 
   public DefaultMacroEvaluator(@Nullable WorkflowToken workflowToken, Map<String, String> runtimeArguments,
-                               long logicalStartTime) {
+                               long logicalStartTime, SecureStore secureStore, String namespace) {
```
this class needs a namespace only for calling secure store get method, would have been nicer if we had `NamespacedSecureStore` implementation. @Nishith ?

gsps1 commented on cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java:
```diff
@@ -73,9 +82,15 @@ public String evaluate(String macroFunction, String... arguments) throws Invalid
+        throw new InvalidMacroException("Secure store macro function only supports 1 argument.");
+      }
+      try {
+        return new String(secureStore.get(namespace, arguments[0]).get());
```
use `Bytes.toString(..)`

albertshau commented on cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/PluginFunctionContext.java:
```diff
@@ -72,10 +74,12 @@ public PluginFunctionContext(String stageName, JavaSparkExecutionContext sec, Pi
 
   public <T> T createPlugin() throws Exception {
-    MacroEvaluator macroEvaluator = new DefaultMacroEvaluator(arguments, logicalStartTime);
+    MacroEvaluator macroEvaluator = new DefaultMacroEvaluator(arguments, logicalStartTime, sec, namespace);
```
it looks like the latest develop has a getSecureStore() method in JavaSparkExecutionContext.  You need to use that method and store just the SecureStore as a field of this class, otherwise Spark serialization/deserialization won't work.

albertshau commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/etl/batch/ETLBatchRuntimeDatasetTest.java:
```diff
@@ -155,4 +156,46 @@ public void testNoMacroMapReduce() throws Exception {
+   * Tests the secure storage macro function by creating dataset names from the secure store data.
+   */
+  @Test
+  public void testSecureStoreMacroMapReduce() throws Exception {
```
lets move this into the DataPipelineTest and have a test case for Spark as well.

knazir commented:
Addressed comments, would appreciate another look through when possible.

albertshau commented on cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java:
```diff
@@ -1207,6 +1211,88 @@ private void testMultipleJoiner(Engine engine) throws Exception {
+   * Tests the secure storage macro function in MapReduce pipelines by creating datasets from the secure store data.
+   */
+  @Test
+  public void testSecureStoreMacroMapReduce() throws Exception {
```
this test is the same as the spark one except for a setEngine() call. Can refactor to have a testSecureMacro() method that takes the engine as an argument, then the two tests just call that.

albertshau commented:
one comment about unit test code reuse. Otherwise lgtm, fix then merge

gsps1 commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/etl/batch/ETLBatchRuntimeDatasetTest.java:
```diff
@@ -29,6 +29,7 @@
 import org.junit.Assert;
 import org.junit.Test;
 
+import java.util.HashMap;
```
unused import ?

gsps1 commented on cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/PluginFunctionContext.java:
```diff
@@ -72,10 +75,12 @@ public PluginFunctionContext(String stageName, JavaSparkExecutionContext sec, Pi
     }
     this.arguments = arguments;
     this.pipelineStr = GSON.toJson(pipelinePhase);
+    this.secureStore = sec.getSecureStore();
```
rather create macro evaluator here, so you don't have to create these variables

albertshau replied:
this class needs to be serializable, so we can't do that unless you make the evaluator serializable too, but even then most of these variables are needed for plugin instantiation

knazir commented:
Build running: http://builds.cask.co/browse/CDAP-DUT4471

knazir commented:
Guild is breen: http://builds.cask.co/browse/CDAP-DUT4471-4 but with the refactoring of the secure store API from @Nishith's PR I'll rebase and fix my changes.

albertshau commented:
lgtm
